src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,635 @@
+    theta += PI;
+    phi += PI;
+    circuit.global_phase -= PI / 2.;
+    // emit circuit
+    pfun(&mut circuit, lam);
+    match xpifun {
+        Some(mut xpifun) if mod_2pi(theta).abs() < atol => xpifun(&mut circuit),
+        _ => {
+            xfun(&mut circuit);
+            pfun(&mut circuit, theta);
+            xfun(&mut circuit);
+        }
+    };
+    pfun(&mut circuit, phi);
+    circuit
+}
+
+fn circuit_rr(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut circuit = Vec::new();
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    if theta.abs() < atol && phi.abs() < atol && lam.abs() < atol {
+        return OneQubitGateSequence {
+            gates: circuit,
+            global_phase: phase,
+        };
+    }
+    if (theta - PI).abs() > atol {
+        circuit.push((String::from("r"), vec![theta - PI, mod_2pi(PI / 2. - lam)]));
+    }
+    circuit.push((String::from("r"), vec![PI, mod_2pi(0.5 * (phi - lam + PI))]));
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+#[pyfunction]
+pub fn generate_circuit(
+    target_basis: &str,
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> PyResult<OneQubitGateSequence> {
+    let res = match target_basis {
+        "ZYZ" => circuit_kak(theta, phi, lam, phase, "rz", "ry", simplify, atol),
+        "ZXZ" => circuit_kak(theta, phi, lam, phase, "rz", "rx", simplify, atol),
+        "XZX" => circuit_kak(theta, phi, lam, phase, "rx", "rz", simplify, atol),
+        "XYX" => circuit_kak(theta, phi, lam, phase, "rx", "ry", simplify, atol),
+        "U3" => circuit_u3(theta, phi, lam, phase, simplify, atol),
+        "U321" => circuit_u321(theta, phi, lam, phase, simplify, atol),
+        "U" => circuit_u(theta, phi, lam, phase, simplify, atol),
+        "PSX" => {
+            let mut inner_atol = match atol {
+                Some(atol) => atol,
+                None => DEFAULT_ATOL,
+            };
+            if !simplify {
+                inner_atol = -1.0;
+            }
+            let fnz = |circuit: &mut OneQubitGateSequence, phi: f64| {
+                let phi = mod_2pi(phi);
+                if phi.abs() > inner_atol {
+                    circuit.gates.push((String::from("p"), vec![phi]));
+                }
+            };
+            let fnx = |circuit: &mut OneQubitGateSequence| {
+                circuit.gates.push((String::from("sx"), Vec::new()));
+            };
+
+            circuit_psx_gen(
+                theta,
+                phi,
+                lam,
+                phase,
+                simplify,
+                atol,
+                fnz,
+                fnx,
+                None::<Box<dyn FnMut(&mut OneQubitGateSequence)>>,
```

Comment:
wow I'm glad this isn't a public-facing API haha

Reply:
Lol, yeah this took me longer to figure out than I wanted. The compiler wasn't able to implicitly figure out the type of the other arm of the `Option<T>` with just a `None` here and `T` it's required to be `Sized` so I had to wrap it a `Box` because `dyn` types aren't sized.

Reply:
Yeah, it all makes sense when I saw what you'd done, but I've never have thought of it without seeing your work!